src/zarr/core/dtype/npy/common.py

```diff
@@ -0,0 +1,583 @@
+    covariant=True,
+)
+TFloatScalar_co = TypeVar(
+    "TFloatScalar_co", bound=np.float16 | np.float32 | np.float64, covariant=True
+)
+
+TComplexDType_co = TypeVar(
+    "TComplexDType_co", bound=np.dtypes.Complex64DType | np.dtypes.Complex128DType, covariant=True
+)
+TComplexScalar_co = TypeVar("TComplexScalar_co", bound=np.complex64 | np.complex128, covariant=True)
+
+
+def endianness_from_numpy_str(endianness: EndiannessNumpy) -> Endianness | None:
+    """
+    Convert a numpy endianness string literal to a human-readable literal value.
+
+    Parameters
+    ----------
+    endianness : Literal[">", "<", "=", "|"]
+        The numpy string representation of the endianness.
+
+    Returns
+    -------
+    Endianness or None
+        The human-readable representation of the endianness.
+
+    Raises
+    ------
+    ValueError
+        If the endianness is invalid.
+    """
+    match endianness:
+        case "=":
+            # Use the local system endianness
+            return sys.byteorder
+        case "<":
+            return "little"
+        case ">":
+            return "big"
+        case "|":
+            # for dtypes without byte ordering semantics
+            return None
+    raise ValueError(
+        f"Invalid endianness: {endianness!r}. Expected one of {get_args(EndiannessNumpy)}"
+    )
+
+
+def endianness_to_numpy_str(endianness: Endianness | None) -> EndiannessNumpy:
+    """
+    Convert an endianness literal to its numpy string representation.
+
+    Parameters
+    ----------
+    endianness : Endianness or None
+        The endianness to convert.
+
+    Returns
+    -------
+    Literal[">", "<", "|"]
+        The numpy string representation of the endianness.
+
+    Raises
+    ------
+    ValueError
+        If the endianness is invalid.
+    """
+    match endianness:
+        case "little":
+            return "<"
+        case "big":
+            return ">"
+        case None:
+            return "|"
+    raise ValueError(
+        f"Invalid endianness: {endianness!r}. Expected one of {get_args(Endianness)} or None"
+    )
+
+
+def float_from_json_v2(data: JSONFloat) -> float:
+    """
+    Convert a JSON float to a float (Zarr v2).
+
+    Parameters
+    ----------
+    data : JSONFloat
+        The JSON float to convert.
+
+    Returns
+    -------
+    float
+        The float value.
+    """
+    match data:
+        case "NaN":
+            return float("nan")
+        case "Infinity":
+            return float("inf")
+        case "-Infinity":
+            return float("-inf")
+        case _:
+            return float(data)
+
+
+def float_from_json_v3(data: JSONFloat) -> float:
+    """
+    Convert a JSON float to a float (v3).
+
+    Parameters
+    ----------
+    data : JSONFloat
+        The JSON float to convert.
+
+    Returns
+    -------
+    float
+        The float value.
+    """
+    # todo: support the v3-specific NaN handling
+    return float_from_json_v2(data)
```

Comment:
Per the spec (https://zarr-specs.readthedocs.io/en/latest/v3/data-types/index.html#permitted-fill-values) this needs to handle the case of an encoded float, for example: `"AAAAAAAA+H8="` (discovered in one of the array tests)

Reply:
good catch!